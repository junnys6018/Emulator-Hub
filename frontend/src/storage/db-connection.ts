import { openDB } from 'idb';
import { EmulatorHubDB, initializeDatabase } from './storage';

const dbConnection = initializeDatabase().then(() => openDB<EmulatorHubDB>('emulator-hub'));

export default dbConnection;
