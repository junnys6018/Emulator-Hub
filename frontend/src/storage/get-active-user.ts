// TODO
/**
 * Returns the uuid of the currently active user
 * if no one is logged in, it is the uuid of the guest account
 * otherwise, it is the uuid of the logged in user
 *
 * Returns `null` if no active user can be found
 */
export default function getActiveUserUuid() {
    return localStorage.getItem('guest-uuid');
}
