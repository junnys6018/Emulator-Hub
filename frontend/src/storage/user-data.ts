import { Record } from './storage';
import { v4 as uuidv4 } from 'uuid';
import { useEffect, useState } from 'react';

export interface UserData extends Record {
    profileImage: Blob;
    userName: string;
}

interface UserProfile {
    profileImage: string;
    userName: string;
}

/**
 * Generate a new guest account with a random color gradient as the profile image
 *
 * May throw
 */
export async function generateGuestAccount(): Promise<UserData> {
    const canvas = document.createElement('canvas');
    canvas.width = 128;
    canvas.height = 128;

    const context = canvas.getContext('2d');
    if (!context) {
        throw new Error('Failed to create canvas context');
    }

    const getRandomColor = () => {
        return `hsl(${Math.random() * 256}, 100%, 50%)`;
    };

    const gradient = context.createLinearGradient(0, 0, 128, 128);
    gradient.addColorStop(0, getRandomColor());
    gradient.addColorStop(1, getRandomColor());

    context.fillStyle = gradient;
    context.fillRect(0, 0, 128, 128);

    const profileImage = await new Promise<Blob>((resolve, reject) => {
        canvas.toBlob(blob => {
            if (blob) {
                resolve(blob);
            } else {
                reject(new Error('Failed to serialize image to a blob'));
            }
        }, 'image/png');
    });

    return {
        uuid: uuidv4(),
        age: 0,
        userName: 'Guest',
        profileImage: profileImage,
    };
}

const whiteImage =
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAIAAAACABAMAAAAxEHz4AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAADUExURf///6fEG8gAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAAfSURBVGje7cExAQAAAMKg9U9tCF8gAAAAAAAAAIBLDSCAAAEf1udwAAAAAElFTkSuQmCC';
export function useUserProfile() {
    const [userProfile, setUserProfile] = useState<UserProfile>({
        profileImage: whiteImage,
        userName: 'Loading...',
    });

    useEffect(() => {
        // setup
        let url = '';
        generateGuestAccount()
            .then(userData => {
                url = URL.createObjectURL(userData.profileImage);
                setUserProfile({
                    profileImage: url,
                    userName: userData.userName,
                });
            })
            .catch(error => {
                alert(error);
            });

        // cleanup
        return () => {
            if (url) {
                URL.revokeObjectURL(url);
            }
        };
    }, []);

    return userProfile;
}
