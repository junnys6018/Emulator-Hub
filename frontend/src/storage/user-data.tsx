import { Record, EmulatorHubDB } from './storage';
import { v4 as uuidv4 } from 'uuid';
import React, { useContext, useEffect, useState } from 'react';
import { openDB } from 'idb';

export interface UserData extends Record {
    profileImage: Blob;
    userName: string;
}

interface UserProfile {
    profileImage: string;
    userName: string;
}

/**
 * Generate a new guest account with a random color gradient as the profile image
 *
 * May throw
 */
export async function generateGuestAccount(): Promise<UserData> {
    const canvas = document.createElement('canvas');
    canvas.width = 128;
    canvas.height = 128;

    const context = canvas.getContext('2d');
    if (!context) {
        throw new Error('Failed to create canvas context');
    }

    const getRandomColor = () => {
        return `hsl(${Math.random() * 256}, 100%, 50%)`;
    };

    const gradient = context.createLinearGradient(0, 0, 128, 128);
    gradient.addColorStop(0, getRandomColor());
    gradient.addColorStop(1, getRandomColor());

    context.fillStyle = gradient;
    context.fillRect(0, 0, 128, 128);

    const profileImage = await new Promise<Blob>((resolve, reject) => {
        canvas.toBlob(blob => {
            if (blob) {
                resolve(blob);
            } else {
                reject(new Error('Failed to serialize image to a blob'));
            }
        }, 'image/png');
    });

    return {
        uuid: uuidv4(),
        age: 0,
        userName: 'Guest',
        profileImage: profileImage,
    };
}

const whiteImage =
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAIAAAACABAMAAAAxEHz4AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAADUExURf///6fEG8gAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAAfSURBVGje7cExAQAAAMKg9U9tCF8gAAAAAAAAAIBLDSCAAAEf1udwAAAAAElFTkSuQmCC';

const UserProfileContext = React.createContext<UserProfile>({
    profileImage: whiteImage,
    userName: 'Loading...',
});

export function UserProfileProvider(props: { children: React.ReactNode }) {
    const [userProfile, setUserProfile] = useState<UserProfile>({
        profileImage: whiteImage,
        userName: 'Loading...',
    });

    useEffect(() => {
        let url = '';
        const guestUuid = localStorage.getItem('guest-uuid');

        const loadUserDataFromUuid = (uuid: string) => {
            openDB<EmulatorHubDB>('emulator-hub', 1)
                .then(db => db.get('users', uuid))
                .then(userData => {
                    if (userData) {
                        url = URL.createObjectURL(userData.profileImage);
                        setUserProfile({
                            profileImage: url,
                            userName: userData.userName,
                        });
                    } else {
                        // TODO
                        alert(`[FATAL] uuid ${uuid} does not point to an account`);
                    }
                });
        };

        if (guestUuid) {
            console.log('[INFO] found guestUuid, loading from storage');
            loadUserDataFromUuid(guestUuid);
        } else {
            console.log('[INFO] guestUuid not found, adding event listener...');
            window.addEventListener('storage', function callback(e) {
                if (e.key === 'guest-uuid' && e.newValue !== null) {
                    console.log('[INFO] guestUuid found, removing event listener...');
                    loadUserDataFromUuid(e.newValue);
                    window.removeEventListener('storage', callback);
                }
            });
        }

        return () => {
            if (url) {
                URL.revokeObjectURL(url);
            }
        };
    }, []);

    return <UserProfileContext.Provider value={userProfile}>{props.children}</UserProfileContext.Provider>;
}

export function useUserProfile() {
    return useContext(UserProfileContext);
}
