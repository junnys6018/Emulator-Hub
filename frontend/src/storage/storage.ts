import { openDB, DBSchema, IDBPDatabase } from 'idb';
import { generateGuestAccount, UserData } from './user-data';
import { v4 as uuidv4 } from 'uuid';

export interface Record {
    uuid: string;
    age: number;
}

export interface EmulatorHubDB extends DBSchema {
    users: {
        value: UserData;
        key: string;
    };
}

export async function createGuestAccount(db: IDBPDatabase<EmulatorHubDB>) {
    console.log('[INFO] No guest account found, generating one');
    const guestUser = await generateGuestAccount();

    const addGuestAccount = async () => {
        try {
            await db.add('users', guestUser);
        } catch (error) {
            if (error.name === 'ConstraintError') {
                console.warn('[WARN] primary key collision, regenerating uuid');
                guestUser.uuid = uuidv4();
                await addGuestAccount();
            } else {
                throw error;
            }
        }
    };
    await addGuestAccount();

    localStorage.setItem('guest-uuid', guestUser.uuid);
    // Dispatch a storage event, this event is not triggered on the window object that modified localStorage
    // so we do it manually
    const event = new StorageEvent('storage', {
        key: 'guest-uuid',
        newValue: guestUser.uuid,
    });
    window.dispatchEvent(event);
}

export async function initializeDatabase() {
    const db = await openDB<EmulatorHubDB>('emulator-hub', 1, {
        upgrade(db, oldVersion) {
            switch (oldVersion) {
                case 0:
                    // database didnt exist, initialize schema
                    console.log('[INFO] Upgrading database to version 1');
                    db.createObjectStore('users', { keyPath: 'uuid' });
            }
        },
        //blocked() {},
        blocking() {
            db.close();
            alert('The application is outdated, please reload the page');
        },
        terminated() {
            alert('[FATAL] database connection was unexpectedly closed');
        },
    });

    // Check if a guest account exists, if not create one
    const guestUuid = localStorage.getItem('guest-uuid');
    if (!guestUuid) {
        await createGuestAccount(db);
    }

    db.close();
}
