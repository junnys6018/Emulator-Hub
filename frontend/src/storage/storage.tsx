import { openDB, DBSchema, IDBPDatabase, OpenDBCallbacks } from 'idb';
import { UserData } from './user-data';
import React, { useContext, useEffect, useState } from 'react';
import { useAlert } from '../components/util/alert';
import { GameData, GameMetaData } from './game-data';
import getActiveUserUuid from './get-active-user';

export interface Record {
    uuid: string;
}

export interface EmulatorHubDB extends DBSchema {
    users: {
        value: UserData;
        key: string;
    };
    gameMetaData: {
        value: GameMetaData;
        key: string;
        indexes: { 'by-user': string };
    };
    gameData: {
        value: GameData;
        key: string;
        indexes: { 'by-user': string };
    };
}

const DatabaseContext = React.createContext<IDBPDatabase<EmulatorHubDB> | null>(null);

export async function initializeDatabase(name: string, options?: OpenDBCallbacks<EmulatorHubDB>) {
    const db = await openDB<EmulatorHubDB>(name, 1, {
        upgrade(db, oldVersion) {
            switch (oldVersion) {
                case 0:
                    // database didnt exist, initialize schema
                    console.log('[INFO] Upgrading database to version 1');
                    db.createObjectStore('users', { keyPath: 'uuid' });

                    const gameMetaData = db.createObjectStore('gameMetaData', { keyPath: 'uuid' });
                    gameMetaData.createIndex('by-user', 'user');

                    const gameData = db.createObjectStore('gameData', { keyPath: 'uuid' });
                    gameData.createIndex('by-user', 'user');
            }
        },
        blocked: options?.blocked,
        blocking() {
            db.close();
            if (options?.blocking) options.blocking();
        },
        terminated: options?.terminated,
    });

    return db;
}

export function DatabaseProvider(props: { children?: React.ReactNode; name?: string }) {
    const [db, setDb] = useState<IDBPDatabase<EmulatorHubDB> | null>(null);
    const alert = useAlert();

    const name = props.name ?? 'emulator-hub';

    useEffect(() => {
        initializeDatabase(name, {
            blocking() {
                alert('The application is outdated, please reload the page', {
                    action: 'REFRESH',
                });
            },
            terminated() {
                alert('The database connection was unexpectedly closed', {
                    severity: 'ERROR',
                    action: 'REFRESH',
                });
            },
        }).then(async db => {
            // Health Check
            const activeUuid = getActiveUserUuid();
            if (activeUuid !== null) {
                const user = await db.get('users', activeUuid);
                if (user === undefined) {
                    // active-uuid points to non-existent user
                    console.log('[WARN] active-uuid did not point to a valid user');
                    localStorage.removeItem('active-uuid');
                }
            }

            setDb(db);
        });
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    return <DatabaseContext.Provider value={db}>{db !== null && props.children}</DatabaseContext.Provider>;
}

export function useDatabase() {
    const db = useContext(DatabaseContext);
    if (db === null) {
        throw new Error('useDatabase must be used within DatabaseProvider');
    }
    return db;
}
