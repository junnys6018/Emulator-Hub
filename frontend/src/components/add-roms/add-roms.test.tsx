/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router';

import AddRoms from './add-roms';

import { useActiveUserProfile } from '@/src/storage/user-data';

jest.mock('../util/message');
jest.mock('../util/alert');
jest.mock('@/src/storage/user-data');

test('<AddRoms /> snapshot', () => {
    (useActiveUserProfile as jest.Mock<any, any>).mockReturnValue([{ userName: 'Jun Lim', profileImage: '/image.png' }]);
    const tree = renderer
        .create(
            <MemoryRouter>
                <AddRoms />
            </MemoryRouter>,
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
