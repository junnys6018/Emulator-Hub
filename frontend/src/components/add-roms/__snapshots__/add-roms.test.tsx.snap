// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddRoms /> snapshot 1`] = `
Array [
  <div
    className="bg-gray-800"
    style={
      Object {
        "backgroundBlendMode": "multiply",
        "backgroundImage": "url(test-file-stub)",
        "backgroundSize": "128px",
      }
    }
  >
    <div
      className="flex flex-wrap"
    >
      <div
        className="container my-7 md:my-10 flex items-center"
      >
        <a
          className="md:hidden font-bold text-2xl xs:text-3xl mr-auto"
          href="/dashboard"
          onClick={[Function]}
        >
          Emulator Hub
        </a>
        <a
          className="hidden md:inline-block font-medium text-xl text-primary-500 hover:underline mr-auto"
          href="/dashboard"
          onClick={[Function]}
        >
          Dashboard
        </a>
        <div
          className="relative dropdown-container hidden md:block"
        >
          <button
            className="select-none h-10 pr-5 rounded-full bg-gray-700 hover:text-green-500"
          >
            <img
              className="inline-block rounded-full mr-2 w-10 h-10 object-cover object-center"
              src="/image.png"
            />
            <span
              className="text-xl align-middle mr-1.5"
            >
              Jun Lim
            </span>
            <svg
              className="inline-block"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 320 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
              />
            </svg>
          </button>
          <div
            className=" pointer-events-none opacity-0 absolute w-max -right-1 transform translate-y-6 z-10 filter drop-shadow bg-gray-700 transition-opacity duration-75 rounded-2xl cursor-default flex flex-col items-stretch dropdown "
            tabIndex={1}
          >
            <img
              className="rounded-full filter drop-shadow my-3 mx-11 object-cover object-center"
              src="/image.png"
              style={
                Object {
                  "height": "120px",
                  "width": "120px",
                }
              }
            />
            <div
              className="border-t border-gray-600 mb-3"
            />
            <a
              className="w-max ml-11 mb-3 text-left hover:text-green-500"
              href="/dashboard"
              onClick={[Function]}
            >
              <svg
                className="inline-block mr-4"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M464 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h416c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM224 416H64v-96h160v96zm0-160H64v-96h160v96zm224 160H288v-96h160v96zm0-160H288v-96h160v96z"
                />
              </svg>
              Dashboard
            </a>
            <a
              className="w-max ml-11 mb-3 text-left hover:text-green-500"
              href="/settings"
              onClick={[Function]}
            >
              <svg
                className="inline-block mr-4"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
                />
              </svg>
              Settings
            </a>
            <a
              className="w-max ml-11 mb-3 text-left hover:text-green-500"
              href="/"
              onClick={[Function]}
            >
              <svg
                className="inline-block mr-4"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M497 273L329 441c-15 15-41 4.5-41-17v-96H152c-13.3 0-24-10.7-24-24v-96c0-13.3 10.7-24 24-24h136V88c0-21.4 25.9-32 41-17l168 168c9.3 9.4 9.3 24.6 0 34zM192 436v-40c0-6.6-5.4-12-12-12H96c-17.7 0-32-14.3-32-32V160c0-17.7 14.3-32 32-32h84c6.6 0 12-5.4 12-12V76c0-6.6-5.4-12-12-12H96c-53 0-96 43-96 96v192c0 53 43 96 96 96h84c6.6 0 12-5.4 12-12z"
                />
              </svg>
              Sign Out
            </a>
            <a
              className="w-max ml-11 mb-3 text-left hover:text-green-500"
              href="/add-roms"
              onClick={[Function]}
            >
              <svg
                className="inline-block mr-4"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                />
              </svg>
              ADD ROMS
            </a>
          </div>
        </div>
        <button
          className="select-none md:hidden"
          onClick={[Function]}
        >
          <img
            className="rounded-full w-10 h-10 object-cover object-center"
            src="/image.png"
          />
        </button>
      </div>
      <div
        className="md:hidden flex-grow w-full bg-gray-900 overflow-hidden transition-max-height max-h-0"
      >
        <div
          className="container flex flex-col"
        >
          <div
            className="flex my-6 items-center"
          >
            <img
              className="rounded-full object-cover object-center"
              src="/image.png"
              style={
                Object {
                  "height": "100px",
                  "width": "100px",
                }
              }
            />
            <span
              className="mx-auto text-4xl font-semibold"
            >
              Jun Lim
            </span>
          </div>
          <a
            className="text-lg w-max active:text-green-500 mb-4"
            href="/dashboard"
            onClick={[Function]}
          >
            <svg
              className="inline-block mr-4"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M464 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h416c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM224 416H64v-96h160v96zm0-160H64v-96h160v96zm224 160H288v-96h160v96zm0-160H288v-96h160v96z"
              />
            </svg>
            Dashboard
          </a>
          <a
            className="text-lg w-max active:text-green-500 mb-4"
            href="/settings"
            onClick={[Function]}
          >
            <svg
              className="inline-block mr-4"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
              />
            </svg>
            Settings
          </a>
          <a
            className="text-lg w-max active:text-green-500 mb-4"
            href="/"
            onClick={[Function]}
          >
            <svg
              className="inline-block mr-4"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M497 273L329 441c-15 15-41 4.5-41-17v-96H152c-13.3 0-24-10.7-24-24v-96c0-13.3 10.7-24 24-24h136V88c0-21.4 25.9-32 41-17l168 168c9.3 9.4 9.3 24.6 0 34zM192 436v-40c0-6.6-5.4-12-12-12H96c-17.7 0-32-14.3-32-32V160c0-17.7 14.3-32 32-32h84c6.6 0 12-5.4 12-12V76c0-6.6-5.4-12-12-12H96c-53 0-96 43-96 96v192c0 53 43 96 96 96h84c6.6 0 12-5.4 12-12z"
              />
            </svg>
            Sign Out
          </a>
          <a
            className="text-lg  tracking-wider w-max active:text-green-500 mb-6"
            href="/add-roms"
            onClick={[Function]}
          >
            <svg
              className="inline-block mr-4"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              />
            </svg>
            ADD ROMS
          </a>
        </div>
      </div>
    </div>
    <div
      className="container flex flex-col pt-2.5 md:pt-0"
    >
      <h2
        className="font-semibold text-xl md:text-3xl"
      >
        Add Roms
      </h2>
      <div
        className="add-roms__dropper border-gray-50"
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
      >
        <SvgrComponent
          className="mb-6 md:transform md:scale-150"
        />
        <span
          className="text-2xl md:hidden"
        >
          Add Roms
        </span>
        <span
          className="text-2xl hidden md:inline-block"
        >
          Drag & Drop Your Roms
        </span>
      </div>
    </div>
  </div>,
  <div
    className="container flex flex-col pt-8 md:pt-16"
  >
    <div
      className="flex mb-4 md:mb-8"
    >
      <h2
        className="font-semibold text-xl md:text-3xl mr-auto"
      >
        New Roms
      </h2>
      <button
        className="btn-primary h-8 md:h-10 px-4 md:px-16 md:text-xl disabled"
        onClick={[Function]}
      >
        Add All
      </button>
    </div>
    <div
      className="add-roms__forms-container"
      id="form-container"
    />
  </div>,
]
`;
